Remove: deprecated native MFA feature

peterhaochen47 opened the pull request:
- Context about its deprecation:
  - This feature is under-utilized, and requires further maintenance for which our team lacks the resource. (For example, this feature is potentially vulnerable because a secure Content-Security-Policy cannot be applied to its pages without breaking them.) The feature has also been marked as "not ready for production" for a few years now. So we opt to remove the feature and instead recommend using the external IDPs's own MFA features. See more context in https://github.com/cloudfoundry/uaa/issues/2196.
- breaking changes planning: https://github.com/cloudfoundry/uaa-release/issues/739
- depends on: https://github.com/cloudfoundry/uaa-release/pull/784

 [#186854489]

peterhaochen47 committed:
Remove: deprecated native MFA feature

- Context about its deprecation:
  - This feature is under-utilized, and requires further
    maintenance for which our team lacks the resource. (For
    example, this feature is potentially vulnerable because
    a secure Content-Security-Policy cannot be applied to its
    pages without breaking them.) The feature has also been
    marked as "not ready for production" for a few years now.
    So we opt to remove the feature and instead recommend
    using the external IDPs's own MFA features. See more context
    in https://github.com/cloudfoundry/uaa/issues/2196.
- This commit removes all MFA-specific codes, except for
  the following, on which we will make follow-up commits:
 - README's deprecation notice
 - database operations
 - Content-Security-Policy's exemption toward MFA endpoint (https://github.com/cloudfoundry/uaa/blob/72565fb56cd1f90af499119d32c891937f3c5a76/server/src/main/java/org/cloudfoundry/identity/uaa/security/web/ContentSecurityPolicyFilter.java#L29)
- breaking changes planning: https://github.com/cloudfoundry/uaa-release/issues/739
- Further notes about specific changes in tests:
  - For PasscodeMockMvcTests.testLoginUsingPasscodeWithUnknownToken(), the assertion
    on response code is changed from 401 to 403. This is because 403 was the original
    asserted value before MFA was added (see: https://github.com/cloudfoundry/uaa/commit/92abee6f58baa4684248ddb4b020426000b5c14a).
    The 403 response also makes sense in the context of the test (authentication
    present but has insufficient access).

 [#186854489]

strehle commented:
@peterhaochen47 I think you can remove
libraries.zxing, https://github.com/cloudfoundry/uaa/blob/develop/dependencies.gradle#L129
libraries.googleAuth, https://github.com/cloudfoundry/uaa/blob/develop/dependencies.gradle#L55

peterhaochen47 committed:
clean up unused MFA-related DB tables

- the MFA feature has been removed, hence
adding DB migrations to drop the feature's
related DB tables.
- also clean up README and tests

[#186854489]

peterhaochen47 committed:
remove unnecessary code

- the MFA feature has been removed (aka the "/login/mfa/" endpoint)
hence the code that creates an exemption for the MFA endpoint when
enforcing "Content-Security-Policy" is no longer needed

[#186854489]

peterhaochen47 committed:
remove: out of date deprecation notice about MFA

- the MFA feature has been removed, so
no longer need the deprecation notice

[#186854489]

peterhaochen47 committed:
remove: unused imports

- the beans are for MFA feature, which
has been removed

[#186854489]

peterhaochen47 committed:
refactor: reduce unnecessary dependency

- use a more standard way to intialize a list,
instead of using com.beust.jcommander.internal.Lists.newArrayList
- motivation: this com.beust.jcommander.internal.Lists.newArrayList
dep was added to our dependencies indirectly via
another dependency I'm trying to remove

[#186854489]

peterhaochen47 committed:
remove: unused dependencies

- these deps are for MFA feature (which
has been removed)
- also: before, the code was using org.apache.httpcomponents:httpclient
but getting it indirectly via the MFA-related
deps (which this commit aims to remove);
hence, now need to directly declare org.apache.httpcomponents:httpclient
as a dependency.

[#186854489]

peterhaochen47 committed:
refactor: simplify code

- by merging cases

strehle commented on server/src/main/java/org/cloudfoundry/identity/uaa/audit/JdbcUnsuccessfulLoginCountingAuditService.java:
```diff
@@ -40,20 +39,11 @@ public JdbcUnsuccessfulLoginCountingAuditService(
-                resetAuthenticationEvents(auditEvent, zoneId, UserAuthenticationFailure);
-                break;
-            case UserAccountUnlockedEvent:
+            case PasswordChangeSuccess, UserAccountUnlockedEvent:
```
see sonar, move this into one list

https://sonarcloud.io/project/issues?resolved=false&sinceLeakPeriod=true&pullRequest=2717&id=cloudfoundry-identity-parent&open=AY2GPGZ4RapHNV97UMiS

peterhaochen47 replied:
Cool, fixed.

strehle reviewed (approved):
ok for us, see sonar comment
